river/covariance/emp.py

```diff
@@ -64,93 +110,99 @@ class CovMatrix:
                 cov = self[i, j]
             cov.update(x[i], x[j])
 
         for i, xi in x.items():
             try:
                 var = self[i, i]
             except KeyError:
-                self._covs[i, i] = stats.Var(self.ddof)
+                self._cov[i, i] = stats.Var(self.ddof)
                 var = self[i, i]
             var.update(xi)
 
         return self
 
-    def update_many(self, X: pd.DataFrame):
-        """Update with many samples.
+    def revert(self, x: dict):
+        """Downdate with a single sample.
 
         Parameters
         ----------
-        X
-            Samples.
+        x
+            A sample.
 
         """
 
-        for i, j in itertools.combinations(sorted(X.columns), r=2):
-            try:
-                cov = self[i, j]
-            except KeyError:
-                self._covs[i, j] = stats.Cov(self.ddof)
-                cov = self[i, j]
-            cov.update_many(X[i].values, X[j].values)
+        for i, j in itertools.combinations(sorted(x), r=2):
+            self[i, j].revert(x[i], x[j])
 
-        for i in X.columns:
-            try:
-                var = self[i, i]
-            except KeyError:
-                self._covs[i, i] = stats.Var(self.ddof)
-                var = self[i, i]
-            var.update_many(X[i].values)
+        for i, xi in x.items():
+            self[i, i].revert(x[i])
 
         return self
 
-    def __getitem__(self, key):
-        """
+    def update_many(self, X: pd.DataFrame):
+        """Update with a dataframe of samples.
 
-        A covariance matrix is symmetric. For ease of use we make the __getitem__ method symmetric.
+        Parameters
+        ----------
+        X
+            A dataframe of samples.
 
         """
-        x, y = key
-        try:
-            return self._covs[x, y]
-        except KeyError:
-            return self._covs[y, x]
 
-    def __repr__(self):
+        # dict -> numpy
+        X_arr = X.values
+        mean_arr = X_arr.mean(axis=0)
+        cov_arr = np.cov(X_arr.T, ddof=self.ddof)
 
-        names = sorted(set(i for i, _ in self._covs))
+        mean = dict(zip(X.columns, mean_arr))
+        cov = {
+            (i, j): cov_arr[r, c]
+            for (r, i), (c, j) in itertools.combinations_with_replacement(enumerate(X.columns), r=2)
+        }
 
-        headers = [""] + list(map(str, names))
-        columns = [headers[1:]]
-        for col in names:
-            column = []
-            for row in names:
-                try:
-                    column.append(f"{self[row, col].get():{self._fmt}}")
-                except KeyError:
-                    column.append("")
-            columns.append(column)
+        for i, j in itertools.combinations(sorted(X.columns), r=2):
+            try:
+                self[i, j]
+            except KeyError:
+                self._cov[i, j] = stats.Cov(self.ddof)
+            self[i, j]._iadd(
+                other_mean_x=mean[i],
```

Comment:
This kind of utility is more valuable than it appears. We might think about creating a standardized way of merging and building new stats objects from precomputed statistics (for instance, initializing an incremental stats estimator from a batch computation).

Reply:
`Var` used to have a `_from_state` method.

In this example, by following the same design pattern, we would do the following:

`other_cov = stats.Cov._from_state(...)`

and then

`self[i, j] += other_cov`

I like this design decision better than having both a `__iadd__` and `_iadd`. It is also generic enough to handle `__sub__` and `__isub__` (or any other stuff we might come up with in the future).

The `_from_state` name is up to debate, of course. We could make almost every stats object have this utility.

Reply:
That's a very good idea! I'll look into it